src/ServerRequestFactory.php

```diff
@@ -99,6 +101,82 @@ public static function fromGlobals(
+            'REQUEST_URI' => $request->server['request_uri'],
+            'PATH_INFO' => $request->server['path_info'],
+            'REQUEST_TIME' => $request->server['request_time'],
+            'GATEWAY_INTERFACE' => 'swoole/' . SWOOLE_VERSION,
+            // Server
+            'SERVER_PROTOCOL' => isset($request->header['server_protocol']) ?
+                                 $request->header['server_protocol'] : $request->server['server_protocol'],
+            'REQUEST_SCHEMA' => isset($request->header['request_scheme']) ?
+                                $request->header['request_scheme'] :
+                                explode('/', $request->server['server_protocol'])[0],
+            'SERVER_NAME' => isset($request->header['server_name']) ?
+                             $request->header['server_name'] : $request->header['host'],
+            'SERVER_ADDR' => $request->header['host'],
+            'SERVER_PORT' => isset($request->header['server_port']) ?
+                             $request->header['server_port'] : $request->server['server_port'],
+            'REMOTE_ADDR' => isset($request->server['remote_addr']) ?
+                             $request->server['remote_addr'] : $request->header['host'],
+            'REMOTE_PORT' => isset($request->header['remote_port']) ?
+                             $request->header['remote_port'] : $request->server['remote_port'],
+            'QUERY_STRING' => isset($request->server['query_string']) ?
+                              $request->server['query_string'] : '',
+            // Headers
+            'HTTP_HOST' => $request->header['host'],
+            'HTTP_USER_AGENT' => isset($request->header['user-agent']) ?
+                                 $request->header['user-agent'] : '',
+            'HTTP_ACCEPT' => isset($request->header['accept']) ?
+                             $request->header['accept'] : '*/*',
+            'HTTP_ACCEPT_LANGUAGE' => isset($request->header['accept-language']) ?
+                                      $request->header['accept-language'] : '',
+            'HTTP_ACCEPT_ENCODING' => isset($request->header['accept-encoding']) ?
+                                      $request->header['accept-encoding'] : '',
+            'HTTP_CONNECTION' => isset($request->header['connection']) ?
+                                 $request->header['connection'] : '',
+            'HTTP_CACHE_CONTROL' => isset($request->header['cache-control']) ?
+                                    $request->header['cache-control'] : '',
+        ];
+
+        $headers = [];
+        foreach ($request->header as $name => $value) {
+            $headers[str_replace('-', '_', $name)] = $value;
+        }
+
+        return new ServerRequest(
+            $server,
+            static::normalizeFiles($files),
+            static::marshalUriFromServer($server, $headers),
+            $server['REQUEST_METHOD'],
+            $request->rawContent(),
```

Comment:
Is there a possibility to retrieve a stream resource, by any chance? If so, we could decorate that in a `Stream` instance, which would be better.

Reply:
It seems that `swoole_http_request` contains only the body message as string.

Reply:
This definitely won't work as-is. `ServerRequestFactory::initialize()` accepts the following types of `$body` arguments:

- A `StreamInterface` instance.
- A string representing a stream URI to open.
- A PHP resource.

Passing just the contents will not work; it needs to be wrapped in a `Stream` instance first.